pkg/parser/parser.go

```diff
@@ -0,0 +1,291 @@
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes"
+
+	apiextensionsv1alpha1 "github.com/crossplane/crossplane/apis/apiextensions/v1alpha1"
+	pkgmetav1alpha1 "github.com/crossplane/crossplane/apis/pkg/meta/v1alpha1"
+)
+
+// Parser is a package parser.
+type Parser interface {
+	Parse(context.Context, ...BackendOption) (*Package, error)
+}
+
+// parser is a Parser implementation with pluggable backends.
+type parser struct {
+	backend Backend
+}
+
+// New returns a new parser.
+func New(backend Backend) Parser {
+	return &parser{
+		backend: backend,
+	}
+}
+
+// BackendOption modifies the parser backend. Backends may accept options at
+// creation time, but must accept them at initialization.
+type BackendOption func(Backend)
+
+// Backend provides a source for the parser.
+type Backend interface {
+	Init(context.Context, ...BackendOption) (io.ReadCloser, error)
+}
+
+// Package includes fields that all meaningful parsers should support.
+type Package struct {
+	provider                     *pkgmetav1alpha1.Provider
+	configuration                *pkgmetav1alpha1.Configuration
+	customResourceDefinitions    map[string]*apiextensions.CustomResourceDefinition
+	compositeResourceDefinitions map[string]*apiextensionsv1alpha1.CompositeResourceDefinition
+	compositions                 map[string]*apiextensionsv1alpha1.Composition
+}
+
+// NewPackage returns a new Package with maps initialized.
+func NewPackage() *Package {
+	return &Package{
+		customResourceDefinitions:    map[string]*apiextensions.CustomResourceDefinition{},
+		compositeResourceDefinitions: map[string]*apiextensionsv1alpha1.CompositeResourceDefinition{},
+		compositions:                 map[string]*apiextensionsv1alpha1.Composition{},
+	}
+}
+
+// GetProvider gets the package provider manifest.
+func (p *Package) GetProvider() *pkgmetav1alpha1.Provider {
+	return p.provider
+}
+
+// GetConfiguration gets the package configuration manifest.
+func (p *Package) GetConfiguration() *pkgmetav1alpha1.Configuration {
+	return p.configuration
+}
+
+// GetCustomResourceDefinitions gets a package's custom resource definitions.
+func (p *Package) GetCustomResourceDefinitions() map[string]*apiextensions.CustomResourceDefinition {
+	return p.customResourceDefinitions
+}
+
+// GetCompositeResourceDefinitions gets a package's composite resource definitions.
+func (p *Package) GetCompositeResourceDefinitions() map[string]*apiextensionsv1alpha1.CompositeResourceDefinition {
+	return p.compositeResourceDefinitions
+}
+
+// GetCompositions gets a package's compositions.
+func (p *Package) GetCompositions() map[string]*apiextensionsv1alpha1.Composition {
+	return p.compositions
+}
+
+// Parse is the underlying parsing logic for parsing Crossplane packages.
+func (p *parser) Parse(ctx context.Context, opts ...BackendOption) (*Package, error) { //nolint:gocyclo
+	pkg := NewPackage()
+	reader, err := p.backend.Init(ctx, opts...)
+	if err != nil || reader == nil {
```

Comment:
It would be an unusual case, the backend init could return `nil` for its `io.ReadCloser`, which could cause a runtime panic if the decoder attempts to read from it. This guards against that.

Reply:
I still don't quite follow. Is there a valid, non-error case in which a backend would return `nil` for its `io.ReadCloser` but also return `nil` for its error? It seems like in that case we'd just return an empty package?

Reply:
Oh sorry I think I misinterpreted what you were saying. Are you just saying that if `reader == nil` then we should return an error?

Reply:
I prefer not to error on nil `io.ReadCloser` to support the `NopBackend` case